src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/MintermClassifier.cs

```diff
@@ -20,81 +20,124 @@ namespace System.Text.RegularExpressions.Symbolic
     /// </remarks>
     internal sealed class MintermClassifier
     {
-        /// <summary>An array used when there's a single minterm, in order to map every ASCII character to it trivially.</summary>
-        private static readonly int[] AllAsciiIsZeroMintermArray = new int[128];
+        /// <summary>An array used to map characters to minterms</summary>
+        private readonly byte[]? _lookup;
 
-        /// <summary>Array providing fast mapping from an ASCII character (the array index) to its corresponding minterm ID.</summary>
-        private readonly int[] _ascii;
-        /// <summary>A multi-terminal BDD for mapping any non-ASCII character to its associated minterm ID.</summary>
-        /// <remarks>
-        /// The use of a multi-terminal BDD here is an implementation detail.  Should we decide its important to optimize non-ASCII inputs further,
-        /// or to consolidate the mechanism with the other engines, an alternatie lookup algorithm / data structure could be employed.
-        /// </remarks>
-        private readonly BDD _nonAscii;
+        /// <summary>
+        /// Fallback lookup if over 255 minterms. This is rarely used.
+        /// </summary>
+        private readonly int[]? _intLookup;
+
+
+        /// <summary>
+        /// Maximum ordinal character for a non-0 minterm, used to conserve memory
+        /// Note: this is maximum index allowed for the lookup, the array size is _maxChar + 1
+        /// </summary>
+        private readonly int _maxChar;
```

Comment:
This is now allocating only up to the highest code point used in the regex, which makes the Kelvin sign cheaper as well.

I'll do some memory testing on some concrete patterns as well but overall i think the memory cost on this side should be very reasonable now.